### PR TITLE
Serial message queue

### PR DESCRIPTION
**Pull request by cait1inhennessy**
While the serial editor is invisible, it saves messages in a queue and displays them upon opening.

Removed all logic surrounding "source" - now graphs are simply separated by variable name.  Separating graphs by source (which is runtime ID for simulator) led to conflict: queued data and live data from the same stream were going to different graphs because the runtime ID changes when you open the serial editor.

Should go into master too.

serial

### CODE REVIEW

File: webapp/src/serial.tsx [hunk 1]
```diff
@@ -14,9 +14,10 @@ import Util = pxt.Util
 const lf = Util.lf
 
 export class Editor extends srceditor.Editor {
+    savedMessageQueue: MessageEvent[] = []
+    maxSavedMessages: number = 10
```
Comment (pelikhan) [inline webapp/src/serial.tsx]: bump it to 100

File: webapp/src/serial.tsx [hunk 2]
```diff
@@ -51,10 +52,12 @@ export class Editor extends srceditor.Editor {
         }
         this.isVisible = b
         if (this.isVisible) {
+            this.processQueuedMessages()
```
Comment (pelikhan) [inline webapp/src/serial.tsx]: I think you want to move this call in ``startRecording``

File: webapp/src/serial.tsx [hunk 3]
```diff
@@ -75,60 +78,76 @@ export class Editor extends srceditor.Editor {
...
-        }
-        let niceSource = this.sourceMap[source]
 
         const m = /^\s*(([^:]+):)?\s*(-?\d+(\.\d*)?)/i.exec(data);
         if (m) {
             const variable = m[2] || '';
             const nvalue = parseFloat(m[3]);
             if (!isNaN(nvalue)) {
-                this.appendGraphEntry(niceSource, variable, nvalue)
+                this.appendGraphEntry(variable, nvalue)
```
Comment (pelikhan) [inline webapp/src/serial.tsx]: are we not differentiating data from different sources? what happens when 2 micro:bit stream the same data?

File: webapp/src/serial.tsx (hunk 2 above)
Comment (cait1inhennessy) [reply webapp/src/serial.tsx]: The editor is always recording when it's opened, so this should be fine as-is.

**Comment by cait1inhennessy** [comment]
@pelikhan This should be ready to merge.

File: webapp/src/serial.tsx [hunk 4]
```diff
@@ -75,27 +79,51 @@ export class Editor extends srceditor.Editor {
...
+        this.savedMessageQueue.forEach(m => this.processMessage(m));
+        this.savedMessageQueue = [];
+    }
+
+    processEvent(ev: MessageEvent) {
+        let msg = ev.data
+        if (msg.type !== "serial") return;
         const smsg = msg as pxsim.SimulatorSerialMessage
+
+        smsg.receivedTime = Util.now();
```
Comment (pelikhan) [inline webapp/src/serial.tsx]: only assign time if missing

**Comment by pelikhan** [comment]
Needs cherry-pick to master.

**Pull request merged**